src/preprocessing/passes/bv_to_int.cpp

```diff
@@ -0,0 +1,1003 @@
+            }
+          }
+          else if (current.isConst())
+          {
+            switch (current.getKind())
+            {
+              case kind::CONST_RATIONAL:
+              {
+                d_bvToIntCache[current] = current;
+                break;
+              }
+              case kind::CONST_BITVECTOR:
+              {
+                // Bit-vector cnostants are transformed into their integer value.
+                BitVector constant(current.getConst<BitVector>());
+                Integer c = constant.toInteger();
+                d_bvToIntCache[current] = d_nm->mkConst<Rational>(c);
+                break;
+              }
+              case kind::CONST_BOOLEAN:
+              {
+                d_bvToIntCache[current] = current;
+                break;
+              }
+              default:
+                throw TypeCheckingException(
+                    current.toExpr(),
+                    string("Cannot translate const: ") + current.toString());
+            }
+          }
+          else
+          {
+            throw TypeCheckingException(
+                current.toExpr(),
+                string("Cannot translate: ") + current.toString());
+          }
+        }
+        else
+        {
+          // The current node has children.
+          // Since we are on the way back up, 
+          // these children were already translated.
+          // We save their translation for future use.
+          vector<Node> intized_children;
+          for (uint64_t i = 0; i < currentNumChildren; i++)
+          {
+            intized_children.push_back(d_bvToIntCache[current[i]]);
+          }
+          // The translation of the current node is determined by the kind of
+          // the node.
+          kind::Kind_t oldKind = current.getKind();
+          switch (oldKind)
+          {
+            case kind::BITVECTOR_PLUS:
+            {
+              uint64_t bvsize = current[0].getType().getBitVectorSize();
+                // we avoid modular arithmetics by the addition of an 
+                // indicator variable sigma.
+                // Tr(a+b) is Tr(a)+Tr(b)-(sigma*2^k), 
+                // with k being the bitwidth,
+                // and sigma being either 0 or 1.
+                Node sigma = d_nm->mkSkolem(
+                    "__bvToInt_sigma_var",
+                    d_nm->integerType(),
+                    "Variable introduced in bvToInt pass to avoid integer mod");
+                Node plus = d_nm->mkNode(kind::PLUS, intized_children);
+                Node multSig = d_nm->mkNode(kind::MULT, sigma, pow2(bvsize));
+                d_bvToIntCache[current] =
+                    d_nm->mkNode(kind::MINUS, plus, multSig);
+                Node zero = d_nm->mkConst<Rational>(0);
+                Node one = d_nm->mkConst<Rational>(1);
+                d_rangeAssertions.insert(d_nm->mkNode(kind::LEQ, zero, sigma));
+                d_rangeAssertions.insert(d_nm->mkNode(kind::LEQ, sigma, one));
```

Comment:
@ahmed-irfan ?

Reply:
I think keeping in the form of conjunctions would be better because it delays the integral reasoning. 
this will first go to the real solver first and if integral reasoning is required, the integer solver will (lazily) add the branch-and-bound clauses. 
Moreover, as @yoni206 said, this appears more consistent with the multiplication case.